### PR TITLE
Checkpoints

### LINKED ISSUE #3: Persistent Headers
We should be saving the last 1152 headers to disk. Flat file storage can work for this. Every hour or so, we can flush the current state to a single file and read it again on boot. It's a couple hundred kb to write every hour, which shouldn't be a big deal.

Having this ready by v1 is ideal, but realistically, this only needs to be ready a couple months after the chain is launched. Headers are fast to sync. It will only become a pain later on.

### LINKED ISSUE #69: Optimize header chain sync using minimal disk space
hnsd currently downloads the entire header chain every time it is launched. While this takes less than one minute currently, it is still an unexpected delay for web-browsing users. https://github.com/handshake-org/hnsd/pull/61 attempts to speed this up by storing all downloaded headers on disk, but hnsd still has to verify each of these from the genesis block to tip on every boot. This also means hnsd will require unlimited disk space as time goes on, consuming an additional  ~12MB per year (headers are 236 bytes * 144 per day * 365 per year).

One thing that is an important difference between hnsd and the Bitcoin SPV model is that a Handshake light name resolver does not need to know every block header in the chain -- we only need to know the current Urkel Tree root! **This means for the purpose of verifying Urkel proofs we only need to store one header: the chain tip.**

However we obviously can't just throw away the entire rest of the header chain -- what if there is a chain reorg? If a reorg happens either during operation or while hnsd is shutdown and then restarted, we need to be able to verify the competing chain from the new tip down to wherever it connects to some other header we've already verified. **This means for the purpose of following the best chain, we need store some amount of historical block headers to stay in sync with the network.**

Solutions:

1. Checkpoints

There is an [open PR](https://github.com/handshake-org/hsd/pull/597) to add checkpoints to hsd. It makes sense to copy these points over to hnsd. At the very least, we don't need to start sync from genesis every boot, we can just start from the highest checkpoint.

2. Chain Locator

Here's where it gets interesting. In Bitcoin and Handshake we already have this data structure called a chain locator. It is a list of some hashes in our best chain that we send to peers when we request blocks. Those peers will respond with blocks in their best chain that connect to one of the hashes in the locator. Usually those new blocks connect directly to our chain tip but sometimes they may connect to an earlier hash if there has been a reorg recently. 

I grabbed a chain locator from a fully synced bcoin node starting up and syncing to peers. These are the chain heights in the locator:

```
[
  697216, 697215, 697214, 697213,
  697212, 697211, 697210, 697209,
  697208, 697207, 697206, 697205,
  697203, 697199, 697191, 697175,
  697143, 697079, 696951, 696695,
  696183, 695159, 693111, 689015,
  680823, 664439, 631671, 566135,
  435063, 172919,      0
] 
```

31 hashes. Note the list is the last 10 blocks, then skip 2, then skip 4, then 8, 16... down to genesis. And this is a locator from a 10-year old chain.

So here's my proposal: hnsd stores enough headers on disk to put together a chain locator, that should be all it needs at all times, after any amount of down time, to reconnect to the network.

### PR DESCRIPTION
**Pull request by pinheadmz**
Closes #3
Closes #69 

### Background

hnsd is a **light name resolver** which is a type of **light client** with NO wallet functions. Since all it does is resolve names, it does NOT require the entire headers chain. In fact, to resolve names all hnsd needs is the `treeRoot` in the header of the chain tip. Like all light clients however, hnsd does require SOME historical consensus data to validate new block headers and stay in sync:

1. The [ONE](https://github.com/handshake-org/hnsd/blob/cf19505f3c0b473e7da800165e42fca36c649019/src/chain.c#L609) previous block header to ensure new headers are in a chain
2. The [ELEVEN](https://github.com/handshake-org/hnsd/blob/cf19505f3c0b473e7da800165e42fca36c649019/src/chain.c#L330) previous block headers to compute median time past
3. The headers from [144 +/-2](https://github.com/handshake-org/hnsd/blob/cf19505f3c0b473e7da800165e42fca36c649019/src/chain.c#L467) blocks ago to compute difficulty target

With this in mind, we introduce a new object called a `checkpoint` which provides enough data for hnsd to effectively begin syncing the chain from ANY arbitrary, non-genesis block. A `checkpoint` contains the following data:

1. Start height `n`
2. The total accumulated work in the blockchain including block at height `n - 1`
3. 150 block headers starting with height `n`, serialized in wire format

With this mechanism in place we can now do some totally badass new things with hnsd:

1. Always start hnsd sync from a specific height (e.g. 136,000) using a hard-coded `checkpoint`.
2. Write our own `checkpoint` to disk during initial chain sync and normal runtime
3. Recover sync status from this saved checkpoint

Overall, this will speed up chain sync after restarts and even for brand new installations tremendously. It will reduce memory usage by hnsd. It is the first time hnsd has ever written anything to disk, and the size of the data written to disk will remain constant (at, like, 36kB).

### Changes

hnsd daemon now accepts two new options:

`-t`, `--checkpoint` (no argument): Begin chain sync at hard-coded checkpoint height of 136,000

`-x <directory path>`, `--prefix <directory path>`: Specify the location on disk to read & write checkpoints

`prefix` will override `checkpoint` if both are used. Recommended usage will be: `hnsd -c -x ~/.hnsd`

When `prefix` is set, hnsd will:

1. Initialize the chain on boot from the saved checkpoint if it exists
2. Save a new checkpoint to disk every 2000 blocks:
  a. Every 2000 blocks defines a "checkpoint window"
  b. At the end of every window, a checkpoint is generated from the **beginning** of that window

Example: At height `8000`, hnsd will store a checkpoint including block headers `6000-6149`. If hnsd is immediately restarted, it will begin syncing from the network with block `6150`. Note that by the time this happens, block `6150` will already have had at least 1,850 confirmations (almost 2 weeks)! This is our safety buffer. A 2000-block reorg would mean the Handshake network is already pretty fucked. If this ever happens, hnsd will "naturally" resync from the genesis block. This is because the genesis block is always included in its chain locator.

### Future work

We can also prune from memory any headers that precede the most recently saved checkpoint. This will limit hnsd memory usage and keep it lean, not included in this PR.

### CODE REVIEW

**Comment by rithvikvibhu** [comment]
I've been running fingertip with `hnsd -t ...` for a few days and everything's good so far. No crashes or problems visiting sites, hnsd syncs < 10 seconds (this is bliss).

**Review by buffrr — COMMENTED** [review]
Great work! This will be a HUGE improvement to hnsd. I will do some more testing if I get the chance but lgtm so far.

File: src/chain.c [hunk 1]
```diff
@@ -312,7 +394,12 @@ hsk_chain_get_locator(const hsk_chain_t *chain, hsk_getheaders_msg_t *msg) {
       height = 0;
 
     hsk_header_t *hdr = hsk_chain_get_by_height(chain, (uint32_t)height);
-    assert(hdr);
+
+    // Due to checkpoint initialization
+    // we may not have any headers from here
+    // down to genesis
+    if (!hdr)
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/chain.c:412]: I added a test that illustrates this now. After syncing 1000 headers, hnsd saves a checkpoint to disk containing headers 800-949. When hnsd restarts the locator it sends to peers is exactly this:

```c
    assert.deepStrictEqual(
      heights,
      [
        949, // tip
             // 10 prev blocks
        948, 947, 946, 945, 944, 943, 942, 941, 940, 939,
        938, // -1
        936, // -2
        932, // -4
        924, // -8
        908, // -16
        876, // -32
        812, // -64
        0    // hnsd doesn't have any blocks lower than 800, so skip to genesis
      ]
    );
```

[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/chain.c:412]: Ok I also just added some more test coverage demonstrating that hnsd can recover, after a restart, from a reorg that forked before, in the middle of, or after the current checkpoint window. The worst case scenario (reorg fork before checkpoint window) results in sync from genesis.

**Comment by rithvikvibhu** [comment]
More of an fyi than a problem, the prefix path not existing is fatal (whereas hsd creates the directory). What do you think about creating the folder if it doesn't exist?

---

>-x <directory path>, --prefix <directory path>: Specify the location on disk to read & write checkpoints
>...Recommended usage will be: hnsd -c -x ~/.hnsd

If only `-x` is specified without a value, how about defaulting to `~/.hnsd` (like hsd does to `~/.hsd`)? But np if no, I understand getting $HOME on all platforms is a pita

**Comment by pinheadmz** [comment]
> More of an fyi than a problem, the prefix path not existing is fatal (whereas hsd creates the directory). What do you think about creating the folder if it doesn't exist?

@rithvikvibhu I'll add this in a future PR. We'll need to implement our own "create directories recursively if they don't yet exist" that works on all platforms. Because this is C and we don't get anything for free!

**Comment by pinheadmz** [comment]
This baby just went through a rigorous rebase. @nodech and @buffrr I explored using libuv for async file i/o and it was too problematic, especially during IBD if we write (and re-write) checkpoint files too quickly. The async becomes a problem because the file writes overlap and need to be atomic, meaning we need a mutex or just... write synchronously. I did clean up a lot of code during that exploration however, and this pull request is now refactored and should be relatively easy to review commit-by-commit.

File: src/daemon.c [hunk 2]
```diff
@@ -547,6 +612,46 @@ int
...
+  }
+
+  if (opt->prefix) {
+    if (!hsk_store_exists(opt->prefix)) {
+      fprintf(stderr, "prefix path does not exist\n");
+      return HSK_EBADARGS;
+    }
+
+    // Prefix must have enough room for filename
+    if (strlen(opt->prefix) + 32 >= HSK_STORE_PATH_MAX) {
```
Comment (nodech) [inline src/daemon.c:632]: Might as well use const for 32 as well or  `hsk_store_filename_len(char *prefix, uint32_t height)`. Define is fine.

Comment (pinheadmz) [reply src/daemon.c]: added in d42ff693566c8d7a56dc7ca44a261199390c2fa3

File: src/store.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+  char buf[HSK_STORE_CHECKPOINT_SIZE];
+  uint8_t *data = (uint8_t *)&buf;
+
+  if (!write_u32be(&data, HSK_MAGIC))
+    goto fail;
+
+  if (!write_u8(&data, HSK_STORE_VERSION))
+    goto fail;
+
+  uint32_t height = chain->height - HSK_STORE_CHECKPOINT_WINDOW;
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/store.c:75]: added in d42ff693566c8d7a56dc7ca44a261199390c2fa3

**Comment by handshake-enthusiast** [comment]
>1. Always start hnsd sync from a specific height (e.g. 136,000) using a hard-coded checkpoint.

What's the logic behind this number? Should it be regenerated from time to time?

When it was first commited in 0f5ef708d141d67c8f3d15092070bcf63b903716 the difference with that moment's height was less than 11k blocks (around 2.5 months). Now it's more than 45k blocks (roughly 8 months).

**Comment by rithvikvibhu** [comment]
The checkpoints in hnsd matches hsd ([which gives ~4 months of PoW](https://github.com/handshake-org/hsd/issues/785#issuecomment-1361832378)). It's part of hsd's [release process](https://github.com/handshake-org/hsd/blob/master/docs/release-process.md#release-candidate-and-major-release) to update checkpoints for every (major) release. hnsd uses the latest checkpoints from hsd, hnsd v2.0.0 was released on Jan 4th this year.

Also fyi, this hard-coded checkpoint only affects _initial_ sync. When using `-x`, hnsd only syncs continuing from last runs. So restarting hnsd would mean almost instant sync (it might re-fetch the last few ones again).

**Comment by handshake-enthusiast** [comment]
@rithvikvibhu thanks for the answer.

>The checkpoints in hnsd matches hsd  
>hnsd uses the latest checkpoints from hsd  

[130000](https://github.com/handshake-org/hsd/blob/v5.0.1/lib/protocol/networks.js#L106) != [136000](https://github.com/handshake-org/hnsd/blob/v2.0.0/src/checkpoints.h#L11)

>this hard-coded checkpoint only affects initial sync..

Yes, this is clear. Still it adds some extra time for the first time that would be nice to decrease. A more recent checkpoint would help. A minor issue.

**Pull request merged**